crates/sparrow-runtime/src/execute/checkpoints.rs

```diff
@@ -0,0 +1,258 @@
+    InvalidResumeFrom(String),
+    #[display(fmt = "invalid 'output_prefix': '{_0}'")]
+    InvalidOutputPrefix(String),
+    #[display(fmt = "expected path '{path:?}' to be in '{prefix:?}'")]
+    InvalidPrefix { prefix: String, path: String },
+    #[display(fmt = "invalid object store")]
+    InvalidObjectStore,
+    #[display(fmt = "listing files in the checkpoint")]
+    ListingFiles,
+    #[display(fmt = "invalid path part '{_0}'")]
+    InvalidPathPart(String),
+}
+
+impl error_stack::Context for Error {}
+
+/// Downloads a compute snapshot from s3 to a local directory.
+pub(crate) async fn download(
+    resume_from: &str,
+    object_stores: &ObjectStoreRegistry,
+    storage_path: &Path,
+    config: &ComputeSnapshotConfig,
+) -> error_stack::Result<(), Error> {
+    let output_prefix = ObjectStoreUrl::from_str(&config.output_prefix)
+        .change_context_lazy(|| Error::InvalidOutputPrefix(config.output_prefix.clone()))?;
+    error_stack::ensure!(
+        output_prefix.is_delimited(),
+        Error::InvalidOutputPrefix(config.output_prefix.clone())
+    );
+    let resume_from = output_prefix
+        .join(resume_from)
+        .change_context_lazy(|| Error::InvalidResumeFrom(resume_from.to_owned()))?;
+
+    let object_store = object_stores
+        .object_store(&resume_from)
+        .change_context(Error::InvalidObjectStore)?;
+
+    let prefix = resume_from.path().change_context(Error::ListingFiles)?;
+    let list_result = object_store
+        .list_with_delimiter(Some(&prefix))
+        .await
+        .into_report()
+        .change_context(Error::ListingFiles)?;
+
+    // Do the downloads, reporting progress (and remaining files).
+    let count = list_result.objects.len();
+    tracing::info!(
+        "Downloading {count} files for checkpoint from {resume_from} to {}",
+        storage_path.display()
+    );
+
+    let resume_from_path = resume_from.path().change_context(Error::DownloadIo)?;
+    futures::stream::iter(list_result.objects)
+        .map(|object| {
+            let Some(relative_path) = object.location.as_ref().strip_prefix(resume_from_path.as_ref()) else {
+                error_stack::bail!(Error::InvalidPrefix {
+                    prefix: resume_from.to_string(),
+                    path: resume_from_path.to_string()
+                })
+            };
+            // Drop the leading `/` (will be left over from the strip prefix above)
+            let relative_path = &relative_path[1..];
+            let source_url = resume_from.join(relative_path).change_context(Error::DownloadIo)?;
+            let destination_path = storage_path.join(relative_path);
+            tracing::info!("For object {object:?}, relative path is {relative_path}, downloading {source_url} to {}", destination_path.display());
+            Ok((source_url, destination_path))
+        })
+        .map_ok(|(source_url, destination_path)| async move {
+            // DO NOT SUBMIT: Can we fix this?
```

Comment:
Is this still a do not submit?

Reply:
Good question. I don't think it needs to be -- it was more a reminder to have this conversation. Specifically, it seems a bit hacky since we need to go back and forth between the `ObjectStoreUrl` (our wrapper of a URL supporting join and such) and the `object_store::path::Path` (which is what it expects and returns).

We *could* add more verbosity / remove some of the safety provided by our wrapper and use the path everywhere. We could also consider wrapping things more deeply... but neither seem *great*.

Looking at this code does it seem like it needs to be fixed now (given we'd need to do one of the above options) or does it seem ok?